lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+
+  void setInstanceLowering(Value oldValue, Value newValue);
+  Value getInstanceLowering(Value oldValue);
+
+  void setSubfieldLowering(Value subfield, Value rootBundle,
+                           std::string suffix);
+  Optional<ValueField> getSubfieldLowering(Value subfield);
+
+  void removeArg(unsigned argNumber);
+  void removeOp(Operation *op);
+  void cleanup();
+
+private:
+  // The builder is set and maintained in the main loop.
+  OpBuilder *builder;
+
+  // State to keep track of arguments and operations to clean up at the end.
+  SmallVector<unsigned, 8> argsToRemove;
+  SmallVector<Operation *, 16> opsToRemove;
+
+  // State to keep a mapping from original bundle values to flattened names and
+  // flattened values.
+  circt::DenseMap<Value, llvm::StringMap<Value>> loweredBundleValues;
+
+  // State to keep a mapping from original instances to new instances with their
+  // bundles flattened.
+  circt::DenseMap<Value, Value> loweredInstances;
+
+  // State to keep a necessary info when traversing potentially nested subfields
+  // of bundles: the root bundle and the fieldname suffix computed so far.
+  circt::DenseMap<Value, ValueField> loweredSubfieldInfo;
+};
+} // end anonymous namespace
+
+/// This is the pass constructor.
+std::unique_ptr<mlir::Pass> circt::firrtl::createLowerFIRRTLTypesPass() {
+  return std::make_unique<FIRRTLTypesLowering>();
+}
+
+// This is the main entrypoint for the lowering pass.
+void FIRRTLTypesLowering::runOnOperation() {
+  auto module = getOperation();
+  auto *body = module.getBodyBlock();
+
+  OpBuilder theBuilder(&getContext());
+  builder = &theBuilder;
+
+  // Lower the module block arguments.
+  SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+  for (auto arg : args)
+    mlir::TypeSwitch<Type>(arg.getType())
+        .Case<BundleType>([&](BundleType type) { lowerArg(arg, type); })
+        .Default<>([&](Type type) { lowerArg(arg, type); });
+
+  // Lower the operations.
+  for (auto &op : body->getOperations()) {
+    builder->setInsertionPoint(&op);
+    if (failed(dispatchVisitor(&op)))
+      return;
+  }
+
+  cleanup();
+}
+
+//===----------------------------------------------------------------------===//
+// Lowering module block arguments.
+//===----------------------------------------------------------------------===//
+
+// Lower arguments with bundle type by flattening them.
+void FIRRTLTypesLowering::lowerArg(BlockArgument arg, BundleType type) {
+  // Flatten the bundle types.
+  SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+  flattenBundleTypes(type, "", false, fieldTypes);
+
+  for (auto field : fieldTypes) {
+    // Create flattened block arguments.
+    auto type = getPortType(field);
+    auto newValue = addArgument(type, arg.getArgNumber(), field.suffix);
+
+    // Map the flattened suffix for the original argument to the new value.
+    setBundleLowering(arg, field.suffix, newValue);
```

Comment:
If the comment above is correct, this just becomes: `if (!field.suffix.empty()) setBundleLowering(..)`

Reply:
I think so... will try this out.